### PR TITLE
Add frame filtering API with Rust core, Python bindings, and tests

### PR DESCRIPTION
**Pull request by leynos**
## Summary
- Introduces a reusable frame filtering API for stack traces and exception payloads, with a Rust core, Python bindings behind the python feature flag, and extensive tests. Includes recursive filtering through causes/contexts and exception groups, plus integration with the logging infrastructure filtering workflow. Also adds design/docs and exposes the new API publicly.

## Changes
- Core filtering utilities
  - Added rust_extension/src/frame_filter.rs with:
    - filter_frames: predicate-based frame filtering
    - limit_frames: keep the N most recent frames
    - exclude_by_filename: exclude frames by filename patterns
    - exclude_by_function: exclude frames by function name patterns
    - exclude_logging_infrastructure: remove common logging frames
    - is_logging_infrastructure: detect logging infra frames
  - Added docs and unit tests for core filtering utilities (Rust)
- Exception/stack payload filtering (recursive)
  - Extended rust_extension/src/exception_schema.rs with:
    - StackTracePayload::filter, ::limit, ::exclude_filenames, ::exclude_functions, ::exclude_logging_infrastructure
    - ExceptionPayload::filter_frames, ::limit_frames, ::exclude_filenames, ::exclude_functions, ::exclude_logging_infrastructure
    - Support for recursive filtering through cause/context and exception groups
- Python bindings
  - Added rust_extension/src/frame_filter_py.rs with Python bindings for:
    - filter_frames(payload, exclude_filenames=None, exclude_functions=None, max_depth=None, exclude_logging=False)
    - get_logging_infrastructure_patterns()
  - Integrated into Rust crate (frame_filter module exposed when Python feature is enabled).
  - Updated rust_extension/src/lib.rs to register Python bindings when the python feature is enabled.
- Public API surface
  - femtologging/__init__.py now exports filter_frames and get_logging_infrastructure_patterns for Python users.
- Tests and BDD coverage
  - New tests in tests/test_frame_filter.py for core filtering logic.
  - New Python tests under tests/ and tests/features/frame_filter.feature with steps in tests/steps/test_frame_filter_steps.py to cover end-to-end usage.
  - Additional unit tests in Rust to exercise new frame_filter and exception filtering paths.
- Documentation
  - docs/frame-filtering-design.md added to describe the API design and filtering flow.

### Python (binding exposed when enabled)
- Filter frames from a payload (stack_info or exc_info)
  - from femtologging import filter_frames, get_logging_infrastructure_patterns
  - payload = {"schema_version": 1, "frames": [{"filename": "myapp/main.py", "lineno": 1, "function": "main"}, {"filename": "femtologging/__init__.py", "lineno": 50, "function": "info"}]}
  - filtered = filter_frames(payload, exclude_logging=True)
  - # filtered contains only non-logging frames

- Get logging infrastructure patterns
  - patterns = get_logging_infrastructure_patterns()
  - # patterns includes substrings like "femtologging", "logging/__init__" etc.

### Rust (library usage)
- StackTracePayload and ExceptionPayload now support:
  - filter_frames(predicate) to filter frames
  - limit_frames(n) to keep the last n frames
  - exclude_filenames(patterns) to drop frames by filename substrings
  - exclude_functions(patterns) to drop frames by function substrings
  - exclude_logging_infrastructure() to drop common logging frames
- Payloads recursively apply filtering across cause, context, and exception groups where applicable

## Test plan
- [x] Rust unit tests for frame_filter utilities
- [x] Rust tests for payload filtering logic (recursion through cause/context/exceptions)
- [x] Python binding tests (filter_frames, get_logging_infrastructure_patterns)
- [x] BDD scenarios validating:
  - Filtering out logging infrastructure from stack payloads
  - Filename-based exclusions
  - Depth limiting
  - Recursive filtering for exception payloads with cause/context/exceptions
  - Combined filtering paths
- [x] End-to-end test coverage via tests/features/frame_filter.feature and tests/steps/test_frame_filter_steps.py

## How to run locally
- Build with Python bindings (if you need Python API):
  - cargo build --features python
  - pytest -q
- Run Rust-only tests: cargo test
- Run full Python integration tests after building with Python bindings: pytest -q

## Notes
- The Python bindings are behind the python feature flag. Enable them when you need Python API access.
- The changes preserve existing payload shapes and add non-breaking extensions to filtering capabilities.
- The new tests exercise edge cases including missing keys, empty frames, and recursive payload structures.

🌿 Generated by [Terry](https://www.terragonlabs.com)

---
ℹ️ Tag @terragon-labs to ask questions and address PR feedback

📎 **Task**: https://www.terragonlabs.com/task/88399e65-5855-4fbf-a85a-7788d53e845c

## Summary by Sourcery

Introduce a reusable frame-filtering API for stack and exception payloads with Rust core utilities, Python bindings, and recursive filtering support.

New Features:
- Add Rust frame filtering utilities for limiting, excluding, and classifying stack frames, including logging infrastructure detection.
- Expose Python bindings and public functions to filter stack/exception payload dicts and retrieve logging infrastructure filename patterns.

Enhancements:
- Extend StackTracePayload and ExceptionPayload to support frame filtering operations, including recursive application across causes, contexts, and exception groups.
- Integrate frame filtering into the Rust extension module exports and Python package namespace.

Documentation:
- Add design documentation describing the frame filtering API, filter flow, and usage examples.

Tests:
- Add Rust unit tests for frame_filter utilities and exception/stack payload filtering behavior.
- Add Python unit tests and BDD scenarios for frame filtering, covering logging exclusion, filename/function filters, depth limiting, recursion, and edge cases.

### CODE REVIEW

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix:

[rust_extension/src/frame_filter_py.rs](https://github.com/leynos/femtologging/pull/305/files/e94195a70c65e1f506083a6113207114e0debb74#diff-830b34c9957dd5aa10eb487161ef294a767d740a6120d32a27200dc1834ac446)

Comment on lines +387 to +414

~~~rust
    fn filter_stack_payload_exclude_logging() {
        Python::with_gil(|py| {
            let payload = make_stack_payload_dict(
                py,
                &[
                    "myapp/main.py",
                    "femtologging/__init__.py",
                    "logging/__init__.py",
                ],
            );

            let result = filter_frames(py, &payload, None, None, None, true).unwrap();
            let result_dict = result.downcast_bound::<PyDict>(py).unwrap();
            let frames = result_dict.get_item("frames").unwrap().unwrap();
            let frames_list = frames.downcast::<PyList>().unwrap();

            assert_eq!(frames_list.len(), 1);
            let frame = frames_list.get_item(0).unwrap();
            let frame_dict = frame.downcast::<PyDict>().unwrap();
            let filename: String = frame_dict
                .get_item("filename")
                .unwrap()
                .unwrap()
                .extract()
                .unwrap();
            assert_eq!(filename, "myapp/main.py");
        });
    }
~~~

❌ New issue: [Code Duplication](https://codescene.io/projects/68445/delta?repo-id=1132401&review-id=305&biomarker=Code+Duplication&filename=rust_extension%2Fsrc%2Fframe_filter_py.rs&method=)
The module contains 2 functions with similar structure: tests.filter_stack_payload_exclude_logging,tests.filter_stack_payload_max_depth

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix:

[rust_extension/src/exception_schema.rs](https://github.com/leynos/femtologging/pull/305/files/a4232a690faf7a22671d97ea32dc32f83a001b21#diff-2ac6c6983994c044cc0666a6bf0427e227b9a5c5599061813fbb5593d2d6288a)

Comment on lines +820 to +851

~~~rust
    fn exception_payload_exclude_filenames_recursive_on_exception_group() {
        let exc1 = ExceptionPayload::new("Error1", "err1").with_frames(vec![
            StackFrame::new("app/module1.py", 1, "func1"),
            StackFrame::new(".venv/lib/foo.py", 2, "venv_func"),
        ]);
        let exc2 = ExceptionPayload::new("Error2", "err2").with_frames(vec![
            StackFrame::new("app/module2.py", 10, "func2"),
            StackFrame::new("site-packages/bar.py", 20, "pkg_func"),
        ]);
        let group = ExceptionPayload {
            schema_version: EXCEPTION_SCHEMA_VERSION,
            type_name: "ExceptionGroup".into(),
            message: "multiple errors".into(),
            exceptions: vec![exc1, exc2],
            frames: vec![StackFrame::new("app/main.py", 100, "main")],
            ..Default::default()
        };

        let filtered = group.exclude_filenames(&[".venv/", "site-packages/"]);

        // Group's own frames unchanged
        assert_eq!(filtered.frames.len(), 1);
        assert_eq!(filtered.frames[0].filename, "app/main.py");

        // exc1's venv frame removed
        assert_eq!(filtered.exceptions[0].frames.len(), 1);
        assert_eq!(filtered.exceptions[0].frames[0].filename, "app/module1.py");

        // exc2's site-packages frame removed
        assert_eq!(filtered.exceptions[1].frames.len(), 1);
        assert_eq!(filtered.exceptions[1].frames[0].filename, "app/module2.py");
    }
~~~

❌ New issue: [Large Assertion Blocks](https://codescene.io/projects/68445/delta?repo-id=1132401&review-id=305&biomarker=Large+Assertion+Blocks&filename=rust_extension%2Fsrc%2Fexception_schema.rs&method=)
The test suite contains 4 assertion blocks with at least 4 assertions, threshold = 4

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?

~~~
Please address the comments from this code review:

## Overall Comments
- The Python-side `filter_stack_payload` rebuilds the payload dict from scratch and only preserves `schema_version` and `frames`; if callers attach additional keys to stack payloads, those will be silently dropped—consider cloning the original dict and updating `frames` in place (similar to how more fields are preserved for exception payloads) to better honor the "preserve payload shapes" goal.
- In `filter_exception_payload`, the `frames` key is omitted entirely when filtering results in an empty list, whereas stack payloads always have a `frames` key; for consistency and simpler consumer logic, consider always setting `frames` (possibly to an empty list) when the input contained that key.

## Individual Comments

### Comment 1
<location> `rust_extension/src/frame_filter_py.rs:149-158` </location>
<code_context>
+}
+
+/// Filter a stack_info payload dict.
+fn filter_stack_payload(
+    py: Python<'_>,
+    payload: &Bound<'_, PyDict>,
+    opts: &FilterOptions<'_>,
+) -> PyResult<PyObject> {
+    let frames = extract_frames(payload)?;
+    let filtered = apply_filters(frames, opts);
+
+    // Build result dict
+    let result = PyDict::new(py);
+
+    // Copy schema_version if present
+    if let Some(version) = payload.get_item("schema_version")? {
+        result.set_item("schema_version", version)?;
+    }
+
+    // Set filtered frames
+    result.set_item("frames", frames_to_py_list(py, &filtered)?)?;
+
+    Ok(result.into())
+}
+
</code_context>

<issue_to_address>
**issue (bug_risk):** Filtering stack payloads drops all fields except `schema_version` and `frames`, which may unexpectedly discard useful data.

This re-creates the payload with only `schema_version` and `frames`, dropping all other keys (e.g. thread info, logger name). If callers rely on those fields surviving frame filtering, this becomes a breaking change. Prefer copying all existing keys and only replacing `frames` (as in `filter_exception_payload`), or clearly define and document which keys must be preserved.
</issue_to_address>

### Comment 2
<location> `tests/test_frame_filter.py:80-89` </location>
<code_context>
+    assert result["frames"][0]["filename"] == "myapp/main.py"
+
+
+def test_stack_exclude_functions() -> None:
+    """Function patterns should exclude matching frames."""
+    payload = make_stack_payload(["a.py", "b.py", "c.py"])
+    payload["frames"][1]["function"] = "_internal_helper"
+
+    result = filter_frames(payload, exclude_functions=["_internal"])
+
+    assert len(result["frames"]) == 2
+    functions = [f["function"] for f in result["frames"]]
+    assert "_internal_helper" not in functions
+
+
</code_context>

<issue_to_address>
**suggestion (testing):** Consider adding a similar function-filtering test for exception payloads, not just stack payloads

`test_stack_exclude_functions` only exercises `exclude_functions` for stack-style payloads. Since exception payloads use a different Rust path (`filter_exception_payload` with recursive filtering), please add a corresponding test (e.g. `test_exc_exclude_functions`) that builds an exception payload, sets one frame’s `function` to match the pattern, and asserts that only that frame is removed while others (including in cause/context or exception groups) remain. This will verify function-name filtering for `exc_info` end-to-end.

Suggested implementation:

```python
    assert len(result["frames"]) == 1
    assert result["frames"][0]["filename"] == "myapp/main.py"


def test_exc_exclude_functions() -> None:
    """Function patterns should exclude matching frames in exception payloads."""
    # Build an exception-style payload with three frames
    exc_payload = make_exc_payload(["a.py", "b.py", "c.py"])

    # Adjust the second frame's function name so it matches the exclude pattern
    # NOTE: the exact path to frames in the exception payload may need to be
    # adapted to your helper's structure (see <additional_changes>).
    frames = exc_payload["exc_info"]["values"][0]["stacktrace"]["frames"]
    frames[1]["function"] = "_internal_helper"

    result = filter_frames(exc_payload, exclude_functions=["_internal"])

    # Re-fetch frames from the filtered payload
    filtered_frames = result["exc_info"]["values"][0]["stacktrace"]["frames"]

    assert len(filtered_frames) == 2
    functions = [f["function"] for f in filtered_frames]
    assert "_internal_helper" not in functions

```

).
    frames = exc_payload["exc_info"]["values"][0]["stacktrace"]["frames"]
    frames[1]["function"] = "_internal_helper"

    result = filter_frames(exc_payload, exclude_functions=["_internal"])

    # Re-fetch frames from the filtered payload
    filtered_frames = result["exc_info"]["values"][0]["stacktrace"]["frames"]

    assert len(filtered_frames) == 2
    functions = [f["function"] for f in filtered_frames]
    assert "_internal_helper" not in functions
>>>>>>> REPLACE
</file_operation>
</file_operations>

<additional_changes>
1. Ensure there is an exception-payload helper named `make_exc_payload` in `tests/test_frame_filter.py` (or imported into it). If it uses a different name or signature, update the test accordingly. For example:
   - If the helper is named `make_exception_payload` or similar, replace `make_exc_payload` with that name.
   - If it takes more complex arguments (e.g. separate frame lists for cause/context/group), construct the main exception's frames as `["a.py", "b.py", "c.py"]` and keep any cause/context/group entries so that this test also exercises recursive filtering.
2. Confirm the exact structure of the exception payload produced by the helper:
   - If frames are not under `["exc_info"]["values"][0]["stacktrace"]["frames"]`, adjust both places in the test where `frames` and `filtered_frames` are obtained to use the correct path (for example, some schemas use `["exception"]["values"]` instead of `["exc_info"]["values"]`).
3. If your existing exception tests already cover nested `cause`, `context`, or exception groups, consider extending this new test to:
   - Set a frame’s `function` inside a nested cause/context/group to `_internal_helper`, and
   - Assert that only those nested frames are removed, while the rest of the structure remains intact.
</issue_to_address>

### Comment 3
<location> `tests/test_frame_filter.py:327-336` </location>
<code_context>
+def test_edge_invalid_frame_missing_filename() -> None:
</code_context>

<issue_to_address>
**suggestion (testing):** Add tests for non-list `frames` and non-dict frame elements to cover remaining type errors in the binding

Current edge tests cover missing keys on frame dicts. The Rust binding also raises when `frames` is not a list or when a frame is not a dict ("'frames' must be a list" / "each frame must be a dict"). Please add tests for these two cases (e.g. `test_edge_frames_not_list` and `test_edge_frame_not_dict`) that pass invalid types and assert the corresponding `TypeError` and message, to fully cover the binding’s error handling for malformed payloads.
</issue_to_address>

### Comment 4
<location> `femtologging/__init__.py:83` </location>
<code_context>
     "basicConfig",
     "dictConfig",
     "fileConfig",
+    "filter_frames",
     "get_logger",
+    "get_logging_infrastructure_patterns",
</code_context>

<issue_to_address>
**issue (review_instructions):** Add behavioural (integration/feature-level) tests for the new frame filtering API; current changes only include unit-style tests in Rust/Python code.

Treat the new `filter_frames` / `get_logging_infrastructure_patterns` API as a feature and cover it with behavioural tests in addition to the existing unit tests in `frame_filter.rs` and `frame_filter_py.rs`. Implement end-to-end tests (e.g. via the public Python API exercising a real log record payload) in `tests/test_frame_filter.py` and populate `tests/features/frame_filter.feature` with scenarios that demonstrate typical usage paths, so the feature is validated at the behavioural level as required by the review instructions.

<details>
<summary>Review instructions:</summary>

**Path patterns:** `**/*`

**Instructions:**
For any new feature or change to an existing feature, both behavioural *and* unit tests are required.

</details>
</issue_to_address>

### Comment 5
<location> `rust_extension/src/frame_filter_py.rs:323` </location>
<code_context>
+        let frames = PyList::empty(py);
+        for (i, filename) in filenames.iter().enumerate() {
+            let frame = PyDict::new(py);
+            frame.set_item("filename", *filename).unwrap();
+            frame.set_item("lineno", i as u32 + 1).unwrap();
+            frame.set_item("function", format!("func_{i}")).unwrap();
</code_context>

<issue_to_address>
**issue (review_instructions):** This test code uses `.unwrap()` instead of `.expect()`, which violates the guideline to prefer `.expect()`.

Per the review instructions, `.unwrap()` should be avoided in favour of `.expect()` so that failures carry clearer messages. Here (and in the other test helpers in this module), please replace `.unwrap()` with `.expect("...")`, e.g.:

```rust
frame.set_item("filename", *filename).expect("failed to set frame filename");
```

The same applies to the other `.unwrap()` calls in this test module (e.g. on `set_item`, `append`, `downcast`, `get_item`, and `extract`).

<details>
<summary>Review instructions:</summary>

**Path patterns:** `**/*.rs`

**Instructions:**
Prefer `.expect()` over `.unwrap()`

</details>
</issue_to_address>

### Comment 6
<location> `docs/frame-filtering-design.md:77` </location>
<code_context>
+- Aligns with the current design where `capture_*` functions return full frames
+- Allows different formatters and handlers to filter differently
+- Maintains testability and composability
+- Keeps worker threads GIL-free (filtering operates on Rust-owned data)
+
+### Filter order
</code_context>

<issue_to_address>
**issue (review_instructions):** The acronym “GIL” is used without being expanded on first use.

To comply with the acronym requirement, expand this as “Global Interpreter Lock (GIL)” the first time it appears.

<details>
<summary>Review instructions:</summary>

**Path patterns:** `**/*.md`

**Instructions:**
Define uncommon acronyms on first use.

</details>
</issue_to_address>
~~~

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix:

[rust_extension/src/exception_schema.rs](https://github.com/leynos/femtologging/pull/305/files/a4232a690faf7a22671d97ea32dc32f83a001b21#diff-2ac6c6983994c044cc0666a6bf0427e227b9a5c5599061813fbb5593d2d6288a)

Comment on lines +820 to +851

~~~rust
    fn exception_payload_exclude_filenames_recursive_on_exception_group() {
        let exc1 = ExceptionPayload::new("Error1", "err1").with_frames(vec![
            StackFrame::new("app/module1.py", 1, "func1"),
            StackFrame::new(".venv/lib/foo.py", 2, "venv_func"),
        ]);
        let exc2 = ExceptionPayload::new("Error2", "err2").with_frames(vec![
            StackFrame::new("app/module2.py", 10, "func2"),
            StackFrame::new("site-packages/bar.py", 20, "pkg_func"),
        ]);
        let group = ExceptionPayload {
            schema_version: EXCEPTION_SCHEMA_VERSION,
            type_name: "ExceptionGroup".into(),
            message: "multiple errors".into(),
            exceptions: vec![exc1, exc2],
            frames: vec![StackFrame::new("app/main.py", 100, "main")],
            ..Default::default()
        };

        let filtered = group.exclude_filenames(&[".venv/", "site-packages/"]);

        // Group's own frames unchanged
        assert_eq!(filtered.frames.len(), 1);
        assert_eq!(filtered.frames[0].filename, "app/main.py");

        // exc1's venv frame removed
        assert_eq!(filtered.exceptions[0].frames.len(), 1);
        assert_eq!(filtered.exceptions[0].frames[0].filename, "app/module1.py");

        // exc2's site-packages frame removed
        assert_eq!(filtered.exceptions[1].frames.len(), 1);
        assert_eq!(filtered.exceptions[1].frames[0].filename, "app/module2.py");
    }
~~~

❌ New issue: [Large Assertion Blocks](https://codescene.io/projects/68445/delta?repo-id=1132401&review-id=305&biomarker=Large+Assertion+Blocks&filename=rust_extension%2Fsrc%2Fexception_schema.rs&method=)
The test suite contains 4 assertion blocks with at least 4 assertions, threshold = 4

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix:

[rust_extension/src/frame_filter/tests.rs](https://github.com/leynos/femtologging/pull/305/files/355711bad040aa9ff5816423370d38d150b5180f#diff-23abb3b0ed9e6ecb28dc6dbb528db0c338abaeeeb62f45b8a174a92c1540a1c2)

Comment on lines +155 to +165

~~~rust
fn exclude_by_function_single_pattern() {
    let frames = vec![
        make_frame("app.py", 1, "main"),
        make_frame("app.py", 2, "_private_helper"),
        make_frame("app.py", 3, "public_api"),
    ];

    let filtered = exclude_by_function(&frames, &["_private"]);

    assert_filter_result_by_function(&filtered, 2, &["main", "public_api"]);
}
~~~

❌ New issue: [Code Duplication](https://codescene.io/projects/68445/delta?repo-id=1132401&review-id=305&biomarker=Code+Duplication&filename=rust_extension%2Fsrc%2Fframe_filter%2Ftests.rs&method=)
The module contains 3 functions with similar structure: exclude_by_function_multiple_patterns,exclude_by_function_no_matches,exclude_by_function_single_pattern

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix:

[rust_extension/src/frame_filter/tests.rs](https://github.com/leynos/femtologging/pull/305/files/355711bad040aa9ff5816423370d38d150b5180f#diff-23abb3b0ed9e6ecb28dc6dbb528db0c338abaeeeb62f45b8a174a92c1540a1c2)

Comment on lines +155 to +165

~~~rust
fn exclude_by_function_single_pattern() {
    let frames = vec![
        make_frame("app.py", 1, "main"),
        make_frame("app.py", 2, "_private_helper"),
        make_frame("app.py", 3, "public_api"),
    ];

    let filtered = exclude_by_function(&frames, &["_private"]);

    assert_filter_result_by_function(&filtered, 2, &["main", "public_api"]);
}
~~~

❌ New issue: [Code Duplication](https://codescene.io/projects/68445/delta?repo-id=1132401&review-id=305&biomarker=Code+Duplication&filename=rust_extension%2Fsrc%2Fframe_filter%2Ftests.rs&method=)
The module contains 3 functions with similar structure: exclude_by_function_multiple_patterns,exclude_by_function_no_matches,exclude_by_function_single_pattern

File: rust_extension/src/frame_filter.rs [hunk 1]
```diff
@@ -0,0 +1,519 @@
...
+
+        // Then exclude virtualenv
+        let step2 = exclude_by_filename(&step1, &[".venv/"]);
+        assert_eq!(step2.len(), 3);
+
+        // Finally limit depth
+        let step3 = limit_frames(&step2, 2);
+        assert_filter_result(&step3, 2, &["myapp/api.py", "myapp/handler.py"]);
+    }
+}
```
[the remark this answers is not in the document]
Comment (leynos) [reply rust_extension/src/frame_filter.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: rust_extension/src/exception_schema/tests.rs [hunk 2]
```diff
@@ -0,0 +1,425 @@
...
+    );
+    assert_eq!(payload.frames.len(), expected_len);
+    for (i, expected) in expected_functions.iter().enumerate() {
+        assert_eq!(
+            payload.frames[i].function, *expected,
+            "Mismatch at frame index {}",
+            i
+        );
+    }
+}
```
[the remark this answers is not in the document]
Comment (leynos) [reply rust_extension/src/exception_schema/tests.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: tests/test_frame_filter.py [hunk 3]
```diff
@@ -0,0 +1,450 @@
...
+    filenames: list[str],
+    type_name: str = "ValueError",
+    message: str = "test error",
+) -> dict:
+    """Create an exc_info payload dict with the given filenames."""
+    payload = make_stack_payload(filenames)
+    payload["type_name"] = type_name
+    payload["message"] = message
+    return payload
+
```
[the remark this answers is not in the document]
Comment (leynos) [reply tests/test_frame_filter.py]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: tests/test_frame_filter.py [hunk 4]
```diff
@@ -0,0 +1,450 @@
...
+    """Max depth should keep only the most recent frames."""
+    payload = make_stack_payload(["a.py", "b.py", "c.py", "d.py", "e.py"])
+
+    result = filter_frames(payload, max_depth=2)
+
+    assert len(result["frames"]) == 2
+    # Should be the last 2 frames (d.py, e.py)
+    assert result["frames"][0]["filename"] == "d.py"
+    assert result["frames"][1]["filename"] == "e.py"
+
```
[the remark this answers is not in the document]
Comment (leynos) [reply tests/test_frame_filter.py]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: tests/test_frame_filter.py [hunk 5]
```diff
@@ -0,0 +1,450 @@
...
+
+def test_frame_locals_preserved() -> None:
+    """Frame locals should be preserved after filtering."""
+    payload = make_stack_payload(["a.py", "b.py"])
+    payload["frames"][0]["locals"] = {"x": "42", "y": "hello"}
+
+    result = filter_frames(payload)
+
+    assert result["frames"][0]["locals"] == {"x": "42", "y": "hello"}
```
[the remark this answers is not in the document]
Comment (leynos) [reply tests/test_frame_filter.py]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai review

File: rust_extension/src/frame_filter_py.rs [hunk 6]
```diff
@@ -0,0 +1,578 @@
...
+        });
+    }
+
+    #[rstest]
+    #[serial]
+    fn get_logging_patterns_returns_expected() {
+        let patterns = get_logging_infrastructure_patterns();
+        assert!(patterns.contains(&"femtologging"));
+        assert!(patterns.contains(&"logging/__init__"));
+    }
```
[the remark this answers is not in the document]
Comment (leynos) [reply rust_extension/src/frame_filter_py.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix:

[tests/steps/test_frame_filter_steps.py](https://github.com/leynos/femtologging/pull/305/files/53f5668cb2172928bad359a4337764ab25ad945c#diff-3d8c6f628ebda321e3cd3e5d47a80d6e566ceedbb789cfbb1ce4a52f37647ded)

Comment on lines +47 to +64

~~~python
def create_stack_payload(filter_fixture: FilterFixture, filenames: str) -> None:
    """Create a stack_info payload from comma-separated filenames.

    Parameters
    ----------
    filter_fixture : FilterFixture
        Shared test state storage.
    filenames : str
        Comma-separated quoted filenames (e.g., '"a.py", "b.py"').

    Returns
    -------
    None
        Stores payload in filter_fixture["payload"].

    """
    parsed = _parse_filenames(filenames)
    filter_fixture["payload"] = make_stack_payload(parsed)
~~~

❌ New issue: [Code Duplication](https://codescene.io/projects/68445/delta?repo-id=1132401&review-id=305&biomarker=Code+Duplication&filename=tests%2Fsteps%2Ftest_frame_filter_steps.py&method=)
The module contains 7 functions with similar structure: check_cause_frame_filename,check_frame_filename,create_exception_payload,create_stack_payload and 3 more functions

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix:

[tests/steps/test_frame_filter_steps.py](https://github.com/leynos/femtologging/pull/305/files/6e3d7d1c7bfafbb351cf1528478c88cfc744ebc7#diff-3d8c6f628ebda321e3cd3e5d47a80d6e566ceedbb789cfbb1ce4a52f37647ded)

Comment on lines +129 to +146

~~~python
def create_stack_payload(filter_fixture: FilterFixture, filenames: str) -> None:
    """Create a stack_info payload from comma-separated filenames.

    Parameters
    ----------
    filter_fixture : FilterFixture
        Shared test state storage.
    filenames : str
        Comma-separated quoted filenames (e.g., '"a.py", "b.py"').

    Returns
    -------
    None
        Stores payload in filter_fixture["payload"].

    """
    parsed = _parse_filenames(filenames)
    filter_fixture["payload"] = make_stack_payload(parsed)
~~~

❌ New issue: [Code Duplication](https://codescene.io/projects/68445/delta?repo-id=1132401&review-id=305&biomarker=Code+Duplication&filename=tests%2Fsteps%2Ftest_frame_filter_steps.py&method=)
The module contains 9 functions with similar structure: check_cause_frame_count,check_cause_frame_filename,check_frame_count,check_frame_filename and 5 more functions

**Comment by leynos** [comment]
@coderabbitai review

File: tests/steps/test_frame_filter_steps.py [hunk 7]
```diff
@@ -0,0 +1,470 @@
...
+    pattern : str
+        Pattern expected to be in the list.
+
+    Returns
+    -------
+    None
+        Raises AssertionError if pattern is not found.
+
+    """
+    patterns = filter_fixture["patterns"]
```
[the remark this answers is not in the document]
Comment (leynos) [reply tests/steps/test_frame_filter_steps.py]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai review

**Comment by leynos** [comment]
@coderabbitai review
